### PR TITLE
New offline DQM module for SMP

### PR DESCRIPTION
**Pull request by anmehta**
#### PR description: this PR includes the implementation of the offline dqm module for  cmssw release validation task for the smp pag. This C++ module includes distributions based on leptons, jets and missing transverse momentum corresponding to Z, W, and VBS-like scenarios. 

the code has been tested locally and the results were presented in the SMP General meetings:
https://indico.cern.ch/event/850145/
https://indico.cern.ch/event/868776/

#### PR validation:
the results have been validated using the drell-yan MC and data rel-val samples from different years of data-taking.

ran the scram build with code checks and format options and the suggested changes are included in the code.

### CODE REVIEW

**Comment by anmehta** [comment]
please test workflow 1330.0,2017.9,2018.9,1320.0

**Comment by jfernan2** [comment]
@anmehta please add a proper title to this PR and description/validation of itself. Thanks

**Comment by anmehta** [comment]
> @anmehta please add a proper title to this PR and description/validation of itself. Thanks

@jfernan2 i've added some description about the PR. sorry, about having missed this earlier.

**Comment by jfernan2** [comment]
Thanks @anmehta can you please add a proper title to PR too, specifying this is SMP and DQM related?

**Comment by anmehta** [comment]
> Thanks @anmehta can you please add a proper title to PR too, specifying this is SMP and DQM related?

@jfernan2, i was not sure if I could do that but now it's done. hopefully this is better. I would like to test workflows before the final review. can i initiate them now ?

**Comment by anmehta** [comment]
@prebello please check.

**Comment by prebello** [comment]
please test workflow 1330.0,2017.9,2018.9,1320.0

Hi @jfernan2 , this is long waited DQM module for the CMSSW release validation of SMP PAG. Please guide @anmehta to properly adjust it for the next releases in DQM. The code is already validated by the SMP group. Not sure if I am still allowed to review PRs :-)

File: DQM/Physics/src/SMPDQM.cc [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+  edm::Handle<edm::View<reco::Vertex>> pvs;
+  if (!iEvent.getByToken(pvs_, pvs)) {
+    return;
+  }
+
+  unsigned int pvMult = 0;
+
+  if (verbose)
+    cout << "check this" << pvs->size() << endl;
```
Comment (fabiocos) [inline DQM/Physics/src/SMPDQM.cc]: @anmehta general comment here and elsewhere: please replace ```cout``` with a suitable choice for the MessageLogger, depending on the severity of the message. ```cout``` is not thread safe

File: DQM/Physics/src/SMPDQM.cc [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    dphi_lepMET->Fill(selected_lep[0].DeltaPhi(imet));
+    mass_lepMET->Fill((selected_lep[0] + imet).M());
+    pt_lepMET->Fill((selected_lep[0] + imet).Pt());
+    if (!selected_recoPFJets.empty()) {
+      dphi_lepjet1->Fill(selected_recoPFJets[0].DeltaPhi(selected_lep[0]));
+    }
+  }  // W case
+
+  else {
+    std::cout << "zero lepton case" << endl;
```
Comment (fabiocos) [inline DQM/Physics/src/SMPDQM.cc]: @anmehta in this case cout is not even protected by the ```verbose``` flag

Comment (anmehta) [reply DQM/Physics/src/SMPDQM.cc]: @fabiocos done. I removed the cout statements altogether.

**Comment by jfernan2** [comment]
please test workflow 1330.0,2017.9,2018.9,1320.0

**Comment by jfernan2** [comment]
@anmehta It seems you have some CMSSW code rule problem still:

Rule1 : https://raw.githubusercontent.com/cms-sw/cmssw/master/Utilities/ReleaseScripts/python/cmsCodeRules/config.py
Search for "using namespace" or "using std::" in header files

/DQM/Physics/src/SMPDQM.h
[78, 79, 80]

**Comment by anmehta** [comment]
@jfernan2, fixed those and running cmsCodeRulesChecker now gives nothing. thanks!

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
+1
@anmehta I have added you as DQM Validation developer since I understand this is the case

**Comment by silviodonato** [comment]
+1

**Pull request merged**